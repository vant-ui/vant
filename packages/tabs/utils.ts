import { raf } from '../utils/dom/raf';

export function scrollLeftTo(el: HTMLElement, to: number, duration: number) {
  let count = 0;
  const from = el.scrollLeft;
  const frames = Math.round((duration * 1000) / 16);

  function animate() {
    el.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      raf(animate);
    }
  }

  animate();
}
