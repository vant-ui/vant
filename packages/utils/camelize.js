const camelizeRE = /-(\w)/g;
export default function(str) {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '');
}
