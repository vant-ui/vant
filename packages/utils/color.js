export const RED = '#f44';
export const BLUE = '#1989fa';
export const GREEN = '#4b0';
export const GRAY_DARK = '#969799';
