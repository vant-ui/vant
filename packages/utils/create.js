/**
 * Create a component with common options
 */
import '../locale';
import Icon from '../icon';
import i18n from '../mixins/i18n';
import install from './install';
import Loading from '../loading';

export default function(sfc) {
  sfc.install = sfc.install || install;
  sfc.mixins = sfc.mixins || [];
  sfc.mixins.push(i18n);
  sfc.components = sfc.components || {};
  sfc.components.icon = Icon;
  sfc.components.loading = Loading;

  return sfc;
};
