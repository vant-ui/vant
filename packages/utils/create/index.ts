import { createBEM } from './bem';
import { createComponent } from './component';
import { createI18N } from './i18n';

type CreateNamespaceReturn = [
  ReturnType<typeof createComponent>,
  ReturnType<typeof createBEM>,
  ReturnType<typeof createI18N>
];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = 'van-' + name;
  return [createComponent(name), createBEM(name), createI18N(name)];
}
