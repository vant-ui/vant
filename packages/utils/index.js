import Vue from 'vue';

export { use, useSlots } from './use';

export const isServer = Vue.prototype.$isServer;

export function noop() {}

export function isDef(value) {
  return value !== undefined && value !== null;
}

export function isObj(x) {
  const type = typeof x;
  return x !== null && (type === 'object' || type === 'function');
}

export function get(object, path) {
  const keys = path.split('.');
  let result = object;

  keys.forEach(key => {
    result = isDef(result[key]) ? result[key] : '';
  });

  return result;
}

const camelizeRE = /-(\w)/g;
export function camelize(str) {
  return str.replace(camelizeRE, (_, c) => c.toUpperCase());
}

export function isAndroid() {
  /* istanbul ignore next */
  return isServer ? false : /android/.test(navigator.userAgent.toLowerCase());
}

export function range(num, min, max) {
  return Math.min(Math.max(num, min), max);
}
