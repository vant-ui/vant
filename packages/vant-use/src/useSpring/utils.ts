import { SpringConfig } from './config';

export const now =
  typeof performance !== 'undefined'
    ? performance.now.bind(performance)
    : Date.now.bind(Date);

export const isArray = Array.isArray.bind(Array);

const reusedTuple: [number, number] = [0, 0];
export function stepper(
  secondPerFrame: number,
  x: number,
  v: number,
  destX: number,
  spring: SpringConfig,
): [number, number] {
  // Spring stiffness, in kg / s^2

  // for animations, destX is really spring length (spring at rest). initial
  // position is considered as the stretched/compressed position of a spring
  const Fspring = -spring.tension * (x - destX);

  // Damping, in kg / s
  const Fdamper = -spring.friction * v;

  const a = (Fspring + Fdamper) / (spring.mass || 1);

  const newV = v + a * secondPerFrame;
  const newX = x + newV * secondPerFrame;

  const precision = spring.precision || 0.01;

  if (Math.abs(newV) < precision && Math.abs(newX - destX) < precision) {
    reusedTuple[0] = destX;
    reusedTuple[1] = 0;
    return reusedTuple;
  }

  reusedTuple[0] = newX;
  reusedTuple[1] = newV;

  return reusedTuple;
}
