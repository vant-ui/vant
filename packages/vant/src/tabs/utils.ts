import { raf, cancelRaf } from '@vant/use';
import { ScrollElement, getScrollTop, setScrollTop } from '../utils';

let scrollLeftToRafId: number;
let scrollTopToRafId: number;

export function scrollLeftTo(
  scroller: HTMLElement,
  to: number,
  duration: number
) {
  cancelRaf(scrollLeftToRafId);

  let count = 0;
  const from = scroller.scrollLeft;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);

  function animate() {
    scroller.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      scrollLeftToRafId = raf(animate);
    }
  }

  animate();
}

export function scrollTopTo(
  scroller: ScrollElement,
  to: number,
  duration: number,
  callback: () => void
) {
  cancelRaf(scrollTopToRafId);

  let current = getScrollTop(scroller);
  const isDown = current < to;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);
  const step = (to - current) / frames;

  function animate() {
    current += step;

    if ((isDown && current > to) || (!isDown && current < to)) {
      current = to;
    }

    setScrollTop(scroller, current);

    if ((isDown && current < to) || (!isDown && current > to)) {
      scrollTopToRafId = raf(animate);
    } else if (callback) {
      scrollTopToRafId = raf(callback as FrameRequestCallback);
    }
  }

  animate();
}
