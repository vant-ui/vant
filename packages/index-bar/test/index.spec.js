import { mount, trigger, triggerDrag } from '../../../test/utils';
import Vue from 'vue';
import IndexBar from '..';
import IndexAnchor from '../../index-anchor';

Vue.use(IndexBar);
Vue.use(IndexAnchor);

function mockScrollIntoView() {
  const fn = jest.fn();
  Element.prototype.scrollIntoView = fn;
  return fn;
}

test('custom anchor text', () => {
  const wrapper = mount({
    template: `
      <van-index-bar>
        <van-index-anchor index="A">Title A</van-index-anchor>
        <van-index-anchor index="B">Title B</van-index-anchor>
      </van-index-bar>
    `
  });

  expect(wrapper).toMatchSnapshot();
});

test('click and scroll to anchor', () => {
  const onSelect = jest.fn();
  const wrapper = mount({
    template: `
      <van-index-bar @select="onSelect">
        <van-index-anchor index="A" />
        <van-index-anchor index="B" />
      </van-index-bar>
    `,
    methods: {
      onSelect
    }
  });

  const fn = mockScrollIntoView();
  const indexes = wrapper.findAll('.van-index-bar__index');
  indexes.at(0).trigger('click');

  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith('A');
});

test('touch and scroll to anchor', () => {
  const onSelect = jest.fn();
  const wrapper = mount({
    template: `
      <van-index-bar @select="onSelect">
        <van-index-anchor index="A" />
        <van-index-anchor index="B" />
        <van-index-anchor index="XXX" />
      </van-index-bar>
    `,
    methods: {
      onSelect
    }
  });

  const fn = mockScrollIntoView();
  const sidebar = wrapper.find('.van-index-bar__sidebar');
  const indexes = wrapper.findAll('.van-index-bar__index');

  document.elementFromPoint = function (x, y) {
    const index = y / 100;

    if (index === 1 || index === 2) {
      return indexes.at(index).element;
    }

    if (index === 3) {
      return {
        dataset: {}
      };
    }
  };

  // horizontal drag
  triggerDrag(sidebar, 100, 0);
  expect(fn).toHaveBeenCalledTimes(0);

  // vertical drag
  trigger(sidebar, 'touchstart', 0, 0);
  trigger(sidebar, 'touchmove', 0, 100);
  trigger(sidebar, 'touchmove', 0, 200);
  trigger(sidebar, 'touchmove', 0, 300);
  trigger(sidebar, 'touchmove', 0, 400);
  trigger(sidebar, 'touchend', 0, 400);
  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith('B');
});
