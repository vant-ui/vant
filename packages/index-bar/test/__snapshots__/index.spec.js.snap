// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom anchor text 1`] = `
<div class="van-index-bar">
  <div class="van-index-bar__sidebar" style="z-index: 1;"><span data-idx="0" data-index="A" class="van-index-bar__index">A</span><span data-idx="1" data-index="B" class="van-index-bar__index">B</span><span data-idx="2" data-index="C" class="van-index-bar__index">C</span><span data-idx="3" data-index="D" class="van-index-bar__index">D</span><span data-idx="4" data-index="E" class="van-index-bar__index">E</span><span data-idx="5" data-index="F" class="van-index-bar__index">F</span><span data-idx="6" data-index="G" class="van-index-bar__index">G</span><span data-idx="7" data-index="H" class="van-index-bar__index">H</span><span data-idx="8" data-index="I" class="van-index-bar__index">I</span><span data-idx="9" data-index="J" class="van-index-bar__index">J</span><span data-idx="10" data-index="K" class="van-index-bar__index">K</span><span data-idx="11" data-index="L" class="van-index-bar__index">L</span><span data-idx="12" data-index="M" class="van-index-bar__index">M</span><span data-idx="13" data-index="N" class="van-index-bar__index">N</span><span data-idx="14" data-index="O" class="van-index-bar__index">O</span><span data-idx="15" data-index="P" class="van-index-bar__index">P</span><span data-idx="16" data-index="Q" class="van-index-bar__index">Q</span><span data-idx="17" data-index="R" class="van-index-bar__index">R</span><span data-idx="18" data-index="S" class="van-index-bar__index">S</span><span data-idx="19" data-index="T" class="van-index-bar__index">T</span><span data-idx="20" data-index="U" class="van-index-bar__index">U</span><span data-idx="21" data-index="V" class="van-index-bar__index">V</span><span data-idx="22" data-index="W" class="van-index-bar__index">W</span><span data-idx="23" data-index="X" class="van-index-bar__index">X</span><span data-idx="24" data-index="Y" class="van-index-bar__index">Y</span><span data-idx="25" data-index="Z" class="van-index-bar__index">Z</span></div>
  <div>
    <div class="van-index-anchor">Title A</div>
  </div>
  <div>
    <div class="van-index-anchor">Title B</div>
  </div>
</div>
`;
